File: leetcode/1879.go
package leetcode

import (
	"math"
	"math/bits"
)

func minimumXORSum(nums1 []int, nums2 []int) int {
	n := len(nums2)
	dp := make([]int, 1<<n)
	dp[0] = 0
	for mask := 1; mask < (1 << n); mask++ {
		dp[mask] = math.MaxInt32
		for j := 0; j < n; j++ {
			if mask&(1<<j) != 0 {
				dp[mask] = min(dp[mask], dp[mask^(1<<j)]+(nums2[j]^nums1[bits.OnesCount(uint(mask))-1]))
			}
		}
	}
	return dp[(1<<n)-1]
}
func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

File: leetcode/1879.cpp
class Solution {
public:
    int minimumXORSum(vector<int>& nums1, vector<int>& nums2) {
        int n = nums2.size();
        vector<int> dp((1 << n), INT_MAX);
        dp[0] = 0;
        for (int mask = 1; mask < (1 << n); ++mask) {
            for (int i = 0; i < n; ++i) {
                if (mask & (1 << i))
                    dp[mask] = min(dp[mask], dp[mask ^ (1 << i)] + (nums2[i] ^ nums1[__builtin_popcount(mask) - 1]));
            }
        }
        return dp[(1 << n) - 1];
    }
};
